azext_iot/iothub/providers/helpers/edge_device_config.py

```diff
@@ -0,0 +1,579 @@
+    # raise error if only key or cert provided
+    if (root_cert_path is not None) ^ (root_key_path is not None):
+        raise RequiredArgumentMissingError(
+            "You must provide a path to both the root cert public and private keys."
+        )
+    # create cert if one isn't provided
+    root_cert = (
+        load_ca_cert_info(root_cert_path, root_key_path, root_cert_password)
+        if all([root_cert_path, root_key_path])
+        else create_v3_self_signed_root_certificate()
+    )
+
+    config = EdgeDevicesConfig(
+        version="1.0",
+        auth_method=(auth_type or DeviceAuthType.shared_private_key.value),
+        default_edge_agent=default_edge_agent,
+        template_config_path=device_config_template,
+        devices=[],
+        root_cert=root_cert,
+    )
+    # Process --device arguments
+    all_devices: Dict[str, Dict[str, str]] = {}
+    for device_input in device_args:
+        # assemble device params from nArgs strings
+        device_dict = assemble_nargs_to_dict(device_input)
+        device_id = device_dict.get("id", None)
+        if not device_id:
+            raise InvalidArgumentValueError(
+                "A device argument is missing required parameter 'id'"
+            )
+        if all_devices.get(device_id, None):
+            raise InvalidArgumentValueError(
+                f"Duplicate deviceId '{device_id}' detected"
+            )
+        all_devices[device_id] = device_dict
+
+    for device_id in all_devices:
+        device_dict = all_devices[device_id]
+        deployment = device_dict.get("deployment", None)
+        if deployment:
+            deployment = try_parse_valid_deployment_config(deployment)
+        parent_id = device_dict.get("parent", None)
+        parent_hostname = None
+        if parent_id:
+            parent = all_devices.get(parent_id, {})
+            parent_hostname = parent.get("hostname", None)
+        hostname = device_dict.get("hostname", None)
+        edge_agent = device_dict.get("edge_agent", None)
+        container_auth_arg = device_dict.get("container_auth", "{}")
+        container_auth_obj = process_json_arg(container_auth_arg)
+        container_auth = (
+            EdgeContainerAuth(
+                serveraddress=container_auth_obj.get("serverAddress", None),
+                username=container_auth_obj.get("username", None),
+                password=container_auth_obj.get("password", None),
+            )
+            if container_auth_obj
+            else None
+        )
+        device_config = EdgeDeviceConfig(
+            device_id=device_id,
+            deployment=deployment,
+            parent_id=parent_id,
+            hostname=hostname,
+            parent_hostname=parent_hostname,
+            edge_agent=edge_agent,
+            container_auth=container_auth,
+        )
+        config.devices.append(device_config)
+    return config
+
+
+def assemble_nargs_to_dict(hash_list: List[str]) -> Dict[str, str]:
```

Comment:
If this is exactly the same as the classmethod in ..../deviceupdate/providers/base (looks like it is) we can plug it in common/utility (TBD future reorganization) and reference it from both places